code/datums/components/pellet_cloud.dm

```diff
@@ -0,0 +1,121 @@
+/datum/component/pellet_cloud
+	var/projectile_type
+	var/num_pellets
+	var/list/pellets = list()
+	var/terminated = 0
+	var/hits = 0
+	var/list/targets_hit = list()
+	var/atom/aim_target
+	var/proj_name
+	var/radius = 4
+	var/holdup = FALSE
+
+/datum/component/pellet_cloud/Initialize(projectile_type, magnitude=5)
+	if(!istype(parent, /obj/item/ammo_casing) && !isgrenade(parent))
+		return COMPONENT_INCOMPATIBLE
+	if(isgrenade(parent))
+		holdup = TRUE
+
+	src.projectile_type = projectile_type
+	if(istype(parent, /obj/item/ammo_casing))
+		num_pellets = magnitude
+	else if(isgrenade(parent))
+		radius = magnitude
+
+	var/obj/projectile/p = projectile_type
+	proj_name = initial(p.name)
+
+/datum/component/pellet_cloud/Destroy(force, silent)
+	if(holdup)
+		return QDEL_HINT_LETMELIVE
+	. = ..()
+
+/datum/component/pellet_cloud/RegisterWithParent()
+	if(istype(parent, /obj/item/ammo_casing))
+		RegisterSignal(parent, COMSIG_PELLET_CLOUD_INIT, .proc/create_pellets)
+	else if(isgrenade(parent))
+		RegisterSignal(parent, COMSIG_GRENADE_PRIME, .proc/circle_pellets)
+
+/datum/component/pellet_cloud/UnregisterFromParent()
+	if(holdup)
+		return QDEL_HINT_LETMELIVE
+	UnregisterSignal(parent, list(COMSIG_PELLET_CLOUD_INIT, COMSIG_GRENADE_PRIME))
+
+//datum/component/pellet_cloud/proc/projectile_hit(atom/fired_from, atom/movable/firer, atom/target, Angle)
+	//do_shrapnel(firer, target)
+
+/datum/component/pellet_cloud/proc/create_pellets(obj/item/ammo_casing/A, target, user, fired_from)
+	aim_target = target
+	for(var/pelleties = 0, pelleties < num_pellets, pelleties++)
+		var/obj/projectile/P = new projectile_type(get_turf(parent))
+		P.original = target
+		P.spread = rand(-40,40)
+		P.fired_from = fired_from
+		P.firer = user
+		P.permutated += user
+		P.suppressed = SUPPRESSED_VERY
+		P.preparePixelProjectile(target, fired_from)
+
+		P.fire()
+
+		pellets += P
+		RegisterSignal(P, COMSIG_PROJECTILE_SELF_ON_HIT, .proc/pellet_hit)
+		RegisterSignal(P, COMSIG_PROJECTILE_SWEET_FA, .proc/pellet_range)
+
+
+/datum/component/pellet_cloud/proc/circle_pellets()
+	if(radius < 1)
+		return
+
+	//var/firer = parent
+	var/atom/target = parent
+	var/list/all_the_turfs_were_gonna_lacerate = RANGE_TURFS(radius, target) - RANGE_TURFS(radius-1, target)
+	num_pellets = all_the_turfs_were_gonna_lacerate.len
+	var/turf/target_turf = get_turf(target)
+	for(var/turf/shootat_turf in all_the_turfs_were_gonna_lacerate)
+		var/obj/projectile/P = new projectile_type(get_turf(target))
+		//Shooting Code:
+		P.range = radius+1
+		P.original = shootat_turf
+		P.fired_from = target
+
+		P.firer = target // don't hit ourself that would be really annoying
+		P.permutated += target // don't hit the target we hit already with the flak
+		P.suppressed = SUPPRESSED_VERY
+
+		P.preparePixelProjectile(shootat_turf, target)
+		RegisterSignal(P, COMSIG_PROJECTILE_SELF_ON_HIT, .proc/pellet_hit)
+		RegisterSignal(P, COMSIG_PROJECTILE_SWEET_FA, .proc/pellet_range)
+		P.fire()
+		pellets += P
```

Comment:
Couldn't `fire()` potentially make the projectile hit a target and get it deleted before running the addition of the object to the list?